scripts/tests/py/metadata.py

```diff
@@ -0,0 +1,223 @@
+                raise Exception(err)
+                        
+            if len(sub_args)==1:
+                run_arg_val=self.env.get(sub_args[0])
+                
+            elif len(sub_args)==2:
+                run_arg_val=self.env.get(sub_args[0]).get(sub_args[1])
+
+            # if a argument has been specified in the comment header
+            # but can't be found in the env file, consider it to be
+            # boolean value.
+            if run_arg_val is None:
+                run_arg_val = True
+
+            metadata_dict[run_arg] = run_arg_val
+
+                
+            
+    def __read_args__(self,run_args_lines: List[str],metadata_dict: Dict[str, str]) -> None:
+        """
+        Goes these each raw line in argument header of a test script and extracts argument
+        values from it.
+
+        Parameters:
+        
+        run_args_lines: List[str]
+          Raw lines in argument header
+
+        metadata_dict: Dict[str, str]
+          Dictionary where the extracted arguments will be stored.
+          This represents the side effect of this function.
+
+        Return:
+        None
+
+        """
+        for run_line in run_args_lines:
+            for run_arg_word in run_line.strip().split():
+                run_arg=run_arg_word.split('/',1)[0]
+                if run_arg in metadata_dict:
+                    metadata_dict[run_arg] = run_arg_word
+
+
+    def parse_script(self, py_script_path: str) -> List[Metadata]:
+        """
+        Parses a script and returns a list of metadata object where
+        each element of that list representing run arguments associated
+        with a particular run.
+
+        Parameter:
+        
+        py_script_path: str
+          path to the python test script
+
+        Return:
+        
+        List[Metadata]
+          List of Metadata object where each Metadata element represents
+          the run arguments associated with a particular run defined in
+          the script file.
+        """
+        
+        runs_def_ptrn=re.compile(r'^\s*#\s*test-runner-runs:\s*(.*)$')
+        args_def_ptrn=re.compile(r'^\s*#\s*test-runner-run/([a-zA-Z0-9_]+):\s*(.*)$')
+
+        runs_arg_lines = {}
+        runs_metadata = []
+        
+    
+        with open(py_script_path, 'r', encoding='utf8') as py_script:
+            for line in py_script.readlines():
+            
+                runs_match = runs_def_ptrn.match(line.strip())
+                args_match = args_def_ptrn.match(line.strip())
+            
+                if runs_match:
+                    for run in runs_match.group(1).strip().split():
+                        runs_arg_lines[run]=[]
+
+                elif args_match:
+                    runs_arg_lines[args_match.group(1)].append(args_match.group(2))
+
+        for run in runs_arg_lines:
+            metadata = Metadata()
+            metadata_dict = vars(metadata)
```

Comment:
this feels off: we lose any typesafety here and start operating on dictionaries. Please update the code to use metadata directly: we should enforce data format and not swith to fully dynamic code.

Somewhere in the codebase there should be a check agains expectet string names and metadata variable names.